Modules/signalmodule.c

```diff
@@ -1877,15 +1872,86 @@ _PyErr_CheckSignalsTstate(PyThreadState *tstate)
     return 0;
 }
 
+/* Subroutine of the PyErr_CheckSignals family:
+   Assert that `tstate` is the current thread's state and it is
+   attached; then proceed to do the work of checking for pending
+   signals.  */
+static int
+check_signals_require_attached(PyThreadState *tstate, bool cycle_collect)
+{
+    _Py_AssertHoldsTstate();
+    _Py_CHECK_EMSCRIPTEN_SIGNALS();
+    return check_signals_attached(tstate, cycle_collect);
+}
 
+/* Declared in pycore_pyerrors.h.  */
+int
+_PyErr_CheckSignalsTstate(PyThreadState *tstate)
+{
+    return check_signals_require_attached(tstate, true);
+}
 
+/* Declared in pycore_pyerrors.h.  */
 int
 _PyErr_CheckSignals(void)
 {
-    PyThreadState *tstate = _PyThreadState_GET();
-    return _PyErr_CheckSignalsTstate(tstate);
+    PyThreadState *tstate = PyGILState_GetThisThreadState();
+    return check_signals_require_attached(tstate, false);
 }
 
+/* Declared in pyerrors.h.  */
+int
+PyErr_CheckSignals(void)
+{
+    PyThreadState *tstate = PyGILState_GetThisThreadState();
+    return check_signals_require_attached(tstate, true);
+}
+
+/* Subroutine of the PyErr_CheckSignals family:
+   Determine whether there is actually any work needing to be done.
+   If so, attach the thread state if necessary, and do that work.
+   The 'tstate' argument must be the thread state (if any) for the
+   current thread, but it does not need to be attached.  */
+static int
+check_signals_detached(PyThreadState *tstate, bool cycle_collect)
+{
+    /* If this thread does not have a thread state at all, then it has
+       never been associated with the Python runtime, so it should not
+       attempt to handle signals or run the cycle collector.  */
+    if (!tstate) {
+        return 0;
+    }
+
+    _Py_CHECK_EMSCRIPTEN_SIGNALS();
+
+    /* Don't acquire the GIL if we don't have anything to do.
+       VERIFYME: I *think* every piece of this expression is safe to
+       execute without holding the GIL and is already sufficiently
+       atomic.  */
+    if ((!_Py_ThreadCanHandleSignals(tstate->interp)
```

Comment:
This is missing a check for the remote debugger interface. Otherwise this will not run the remote debugger code when users call this function without the GIL

Reply:
The checks are the same ones that `_PyRunRemoteDebugger` do before entering the actual code